pkg/event-consumers/nats/nats-consumer.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright (c) 2016-2017 Bitnami
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nats
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/kubeless/kubeless/pkg/utils"
+	"github.com/nats-io/go-nats"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	stopM     map[string](chan struct{})
+	stoppedM  map[string](chan struct{})
+	consumerM map[string]bool
```

Comment:
Just wondering but is this map ok in terms of concurrent access? It seems would be used when either `syncNATSTrigger` and `FunctionAddedDeletedUpdated` are called but not sure if they would both happen concurrently.

Reply:
Added locking there is a remote chance of race condition.